### PR TITLE
223 add atlas navigation

### LINKED ISSUE #202: Add Atlas exploration to simulator
- add function that calculates target position dotbot should explore next (cell to explore).
- this cell is a frontier cell that is closest to starting point and to robot position

### LINKED ISSUE #223: Add atlas navigation
A* for path to target calculation
input: start cell coordinate (startX, startY) and stop cell coordinates (stopX, stopY)
output: array of tuples representing a path of cells.
calculate trajectory heading and timeout duration till next turn

### PR DESCRIPTION
**Pull request by Raa37**

### CODE REVIEW

File: DotBot.py [hunk 1]
```diff
@@ -9,43 +9,47 @@
...
-        self.nextBumpY            = None
-        self.nextBumpTime         = None  # time at which DotBot will bump
+        self.nextBumpX          = None  # coordinate the DotBot will bump into next
+        self.nextBumpY          = None
+        self.nextBumpTime       = None  # time at which DotBot will bump
         # is dotBot a relay
-        self.isRelay              = False
+        self.isRelay            = False
+        # if dotbot has bumped
+        self.bumped             = False
```
Comment (twatteyne) [inline DotBot.py]: I don't understand why you need this. When a DotBot bumps, it sends a message, no need to have a long-term class-level variable

File: DotBot.py [hunk 2]
```diff
@@ -56,22 +60,29 @@ def receive(self, frame):
...
+        # set stop time based on timeout if given
+        timeout            = frame['movements'][self.dotBotId]['timeout']
+        stopTime           = timeout + self.simEngine.currentTime() if timeout else math.inf
 
         # set relay status
-        self.isRelay             = frame['movements'][self.dotBotId]['isRelay']
+        self.isRelay       = frame['movements'][self.dotBotId]['isRelay']
 
         # if dotBot is set as relay don't adjust movement
         if self.isRelay:
```
Comment (twatteyne) [inline DotBot.py]: relays can move, remove this paragraph

File: DotBot.py [hunk 3]
```diff
@@ -148,9 +165,31 @@ def _bumpSensorCb(self):
...
         log.debug(f'DotBot {self.dotBotId} stopped at ({self.x}, {self.y}) at {self.simEngine.currentTime()}')
-        log.debug(f'DotBot {self.dotBotId} expected position at ({self.nextBumpX}, {self.nextBumpY}) at {self.nextBumpTime}')
-        assert self.x == self.nextBumpX
-        assert self.y == self.nextBumpY
+        log.debug(
+            f'DotBot {self.dotBotId} expected position at ({self.nextBumpX}, {self.nextBumpY}) at {self.nextBumpTime}'
+        )
+
+        # dotBot bumped
+        self.bumped = True
```
Comment (twatteyne) [inline DotBot.py]: remove, I don't see why needed

File: DotBot.py [hunk 4]
```diff
@@ -175,35 +214,36 @@ def _transmit(self):
             'source':             self.dotBotId,
             'movementDuration':   self.tsMovementStop - self.tsMovementStart,
             'seqNumNotification': self.seqNumNotification,
+            'bumped':             self.bumped
```
Comment (twatteyne) [inline DotBot.py]: I see why you do this.
- Change item name to `hasJustedBumped`
- add a parameter `hasJustedBumped` to `_stopAndTransmit()`, set to `True` or `False` based on where you call it from

File: DotBot.py [hunk 5]
```diff
@@ -318,18 +357,18 @@ def _computeNextBump(self, currentX, currentY, heading, speed, obstacles):
...
-            timetobump     = u.distance((currentX, currentY), (bumpX, bumpY)) / speed
+            timetobump = u.distance((currentX, currentY), (bumpX, bumpY)) / speed
 
-            assert bumpX  >= 0 and bumpY >= 0
+            assert bumpX >= 0 and bumpY >= 0
 
             # round
-            bumpX          = round(bumpX, 3)
-            bumpY          = round(bumpY, 3)
+            bumpX = round(bumpX, 5)
```
Comment (twatteyne) [inline DotBot.py]: why the change?

File: Orchestrator.py [hunk 6]
```diff
@@ -11,88 +11,101 @@
...
+        self.cellsObstacle     = []
+        self.cellsFrontier     = []
+        self.x                 = self.initX
+        self.y                 = self.initY
+        self.dotBotId          = 0
         # for wireless to identify if this is a relay device or not
-        self.isRelay            = False
+        self.isRelay           = False
+        # to know if A* should be used for finding alternative path
+        self.isDirectPathOpen  = True
```
Comment (twatteyne) [inline Orchestrator.py]: what is a "direct open path"?

File: Orchestrator.py [hunk 7]
```diff
@@ -11,88 +11,101 @@
...
-                   'seqNumNotification': None,
+                    'seqNumCommand':        0,
+                    'seqNumNotification':   None,
                     # if dotBot is relay or not
-                   'isRelay':            False,
+                    'isRelay':              False,
+                    # duration of movement until robot stops
+                    'timeout':              None,
+                    # if dotBot bumped last time it stopped
+                    'bumped':               False,
```
Comment (twatteyne) [inline Orchestrator.py]: I don't think the orchestrator needs to know long term this. Please remove.

File: Orchestrator.py [hunk 8]
```diff
@@ -11,88 +11,101 @@
...
+                    # frontier cell dotBot has been given to explore and path to it
+                    'targetCell':           None,
+                    'currentPath':          [],
```
Comment (twatteyne) [inline Orchestrator.py:71]: what is `currentPath` used for? add comment

File: Orchestrator.py [hunk 9]
```diff
@@ -159,15 +163,25 @@ def receive(self,frame):
...
         # update explored cells
-        cellsExplored          = self._computeCellsExplored(dotBot['x'], dotBot['y'], newX, newY)
-        self.cellsExplored    += cellsExplored['cellsExplored']
+        cellsExplored       = self._computeCellsExplored(dotBot['x'], dotBot['y'], newX, newY)
+        self.cellsExplored += cellsExplored['cellsExplored']
 
         # update obstacle cells
-        if cellsExplored['nextCell']:
-            self.cellsObstacle  += [cellsExplored['nextCell']]
+        if cellsExplored['nextCell'] and frame['bumped']:
```
Comment (twatteyne) [inline Orchestrator.py]: reverse the check, for clarity: `if frame['bumped'] and cellsExplored['nextCell']:` (you should change the name "bumped", see above)

File: Orchestrator.py [hunk 10]
```diff
@@ -159,15 +163,25 @@ def receive(self,frame):
...
+            self.cellsObstacle += [cellsExplored['nextCell']]
+
+        if (
+            frame['bumped']                                    and
+            not cellsExplored['cellsExplored']                 and
```
Comment (twatteyne) [inline Orchestrator.py]: add parenthesis for clarity

File: Orchestrator.py [hunk 11]
```diff
@@ -159,15 +163,25 @@ def receive(self,frame):
...
+            dotBot['currentPath']                              and
+            dotBot['currentPath'][0] == dotBot['targetCell']   and
+            dotBot['targetCell'] in self.cellsFrontier
```
Comment (twatteyne) [inline Orchestrator.py:182]: add parenthesis for clarity

File: Orchestrator.py [hunk 12]
```diff
@@ -191,25 +205,81 @@ def receive(self,frame):
...
+        dotBot['bumped'] = frame['bumped']
+
+        log.debug(f'dotbot {dotBot} is at ({newX},{newY})')
+
+        # if dotBot bumped
+        if frame['bumped']:
+            self.isDirectPathOpen = False
+
+        # assign target cell (destination)
+        if not dotBot['targetCell'] or dotBot['targetCell'] not in self.cellsFrontier:
```
Comment (twatteyne) [inline Orchestrator.py]: add parenthesis `if (not dotBot['targetCell']) or (dotBot['targetCell'] not in self.cellsFrontier):`

File: Orchestrator.py [hunk 13]
```diff
@@ -191,25 +205,81 @@ def receive(self,frame):
...
+
+        else:
+            # target cell not explored yet, keep moving towards it
+            targetCell           = dotBot['targetCell']
+
+        # find path to target
 
-        # pick new speed and heading for dotBot
-        (heading, speed)          = self._pickNewMovement(frame['source'])
+        if not cellsExplored['cellsExplored']:
```
Comment (twatteyne) [inline Orchestrator.py]: I don't understand how there is a "cellsExplored" item in variable "cellsExplored"

File: Orchestrator.py [hunk 14]
```diff
@@ -191,25 +205,81 @@ def receive(self,frame):
...
+            startCell           = (self._xy2cell(dotBot['x'], dotBot['y']))
+        else:
+            startCell           = cellsExplored['cellsExplored'][-1]
+
+        if self.isDirectPathOpen == False or not targetCell:
```
Comment (twatteyne) [inline Orchestrator.py]: add parenthesis

File: Orchestrator.py [hunk 15]
```diff
@@ -232,69 +302,70 @@ def _computeCellsExplored(self, startX, startY, stopX, stopY):
...
         # 1 - for when we calculate the next cell beyond the trajectory
-        maxNumCells = u.distance((startX, startY), (stopX, stopY))*2 + 2
+        maxNumCells = u.distance((startX, startY), (stopX, stopY)) * 2 + 2
 
         # movement is not on boundaries
-        if not(
-            (startX == stopX) and (startX == cx or startX == cx+self.MINFEATURESIZE/2) or
-            (startY == stopY) and (startY == cy or startX == cy+self.MINFEATURESIZE/2)
+        if not (
+                (startX == stopX) and (startX == cx or startX == cx + self.MINFEATURESIZE / 2) or
```
Comment (twatteyne) [inline Orchestrator.py]: add parenthesis around groups of or/and

File: Orchestrator.py [hunk 16]
```diff
@@ -495,3 +565,159 @@ def _isCornerFrontier(self, cell):
...
+                closestFrontiersToDotBot  = [cell for (cell, distance) in cellsAndDistancesToDotBot if
+                                            distance == cellsAndDistancesToDotBot[0][1]]
+                targetFrontier            = random.choice(closestFrontiersToDotBot)
+                self.assignedFrontiers   += [targetFrontier]
+
+        return targetFrontier
+
+    # === Navigation
+
+    def _computePath(self, startCell, targetCell, excludeDiagonalCells=False):
```
Comment (twatteyne) [inline Orchestrator.py:601]: I don't have time to read in detail. Add a comment which shows an example input and output and 1 sentence which describes how this function works

File: Utils.py [hunk 17]
```diff
@@ -55,4 +55,16 @@ def setLoggerUname(uname):
     )
     logging.config.dictConfig(LoggingConfig.LOGGINGCONFIG)
 
+# ============== A* helpers
```
Comment (twatteyne) [inline Utils.py:58]: empty newline after

**Review by twatteyne — CHANGES_REQUESTED** [review]
This is very clean code, congrats. I made some suggestions to remove the bumped attribute of DotBot, and rename a couple of things. I don't have time to carefully read everything. I'd like you to indicate what tests you've run confirming all works. I see you added additiona unit tests, congrats.

File: DotBot.py (hunk 1 above)
Comment (Raa37) [reply DotBot.py:52]: because we stop when we reach target too (even if no bump), so we call transmit from multiple functions now with different bump status, sometimes its true, sometimes false

File: DotBot.py (hunk 4 above)
Comment (Raa37) [reply DotBot.py]: but when the sim engine calls this how does it feed in the proper parameters as we just give it the name of the function and it calles it. or do I modify sim engine to also take parameters as inputs?

File: Orchestrator.py (hunk 6 above)
Comment (Raa37) [reply Orchestrator.py]: going from start to stop without obstacles in between (an obstacle free path). should I rename to obstacleFreePath?

File: Orchestrator.py (hunk 13 above)
Comment (Raa37) [reply Orchestrator.py]: it computes cells explored and next cell that would be explored had robot kept moving in same direction. should i rename function to computeCellsExploredAndNextCell()?
